### PR TITLE
[Bug fix] bind_new_parameters for LinearCombination

### PR DESCRIPTION
**Pull request by Qottmann**
Fix problem from discussion in https://github.com/PennyLaneAI/pennylane/pull/5216#discussion_r1532371621

[sc-59515]

### CODE REVIEW

File: pennylane/ops/functions/bind_new_parameters.py [hunk 1]
```diff
@@ -111,6 +111,28 @@ def bind_new_parameters_composite_op(op: CompositeOp, params: Sequence[TensorLik
...
+):
+    new_operands = []
+
+    for operand in op.operands:
+        op_num_params = operand.num_params
+        sub_params = params[:op_num_params]
+        params = params[op_num_params:]
+        new_operands.append(bind_new_parameters(operand, sub_params))
+
+    coeffs, ops = qml.sum(*new_operands).terms()
```
Comment (Qottmann) [inline pennylane/ops/functions/bind_new_parameters.py:126]: Not the prettiest solution but it _works_ for the moment? 🤷

File: pennylane/ops/functions/bind_new_parameters.py [hunk 2]
```diff
@@ -111,6 +111,30 @@ def bind_new_parameters_composite_op(op: CompositeOp, params: Sequence[TensorLik
...
+):
+
+    new_coeffs, new_ops = [], []
+    i = 0
+    for o in op.ops:
+        new_coeffs.append(params[i])
+        i += 1
+        if isinstance(o, qml.Hermitian):
+            new_ops.append(qml.Hermitian(params[i], wires=o.wires))
+            i += 1
```
Comment (mudit2812) [inline pennylane/ops/functions/bind_new_parameters.py:126]:
```suggestion
        if o.data:
            sub_data = params[i : i + len(o.data)]
            new_ops.append(bind_new_parameters(o, sub_data))
            i += len(sub_data)
```

File: pennylane/ops/functions/bind_new_parameters.py [hunk 3]
```diff
@@ -111,6 +111,30 @@ def bind_new_parameters_composite_op(op: CompositeOp, params: Sequence[TensorLik
...
+    op: qml.ops.LinearCombination, params: Sequence[TensorLike]
+):
+
+    new_coeffs, new_ops = [], []
+    i = 0
+    for o in op.ops:
+        new_coeffs.append(params[i])
+        i += 1
+        if isinstance(o, qml.Hermitian):
+            new_ops.append(qml.Hermitian(params[i], wires=o.wires))
```
Comment (astralcai) [inline pennylane/ops/functions/bind_new_parameters.py:125]: I have a question, are we sure that the `params` list is in the order assumed here? The coefficient for the `Hermitian` would come before the data of the `Hermitian` itself in the list of `params`?

Comment (Qottmann) [reply pennylane/ops/functions/bind_new_parameters.py:125]: I am assuming the order of how also `op.data` is generated:
```python
op = qml.ops.LinearCombination([0.5, 1.5], [qml.Hermitian(np.eye(2), 0), X(0)])
op.data
```

```
(0.5,
 array([[1., 0.],
        [0., 1.]]),
 1.5)
```

which seems to follow `params = [c0, op0.data, c1, op1.data,...]` where the `opX.data` is optional depending on whether the op carries data or not

File: tests/ops/functions/test_bind_new_parameters.py [hunk 4]
```diff
@@ -79,6 +80,11 @@
                 qml.sum(qml.GellMann(1, 3), qml.THermitian(GELL_MANN[6], 1)),
             ),
         ),
+        (
+            qml.ops.LinearCombination([2.0, 3.0], [qml.Hermitian(np.eye(2), 0), qml.X(0)]),
+            (1.0, -np.eye(2), -1.0),
+            qml.ops.LinearCombination([1.0, -1.0], [qml.Hermitian(-np.eye(2), 0), qml.X(0)]),
```
Comment (mudit2812) [inline tests/ops/functions/test_bind_new_parameters.py:86]: Can you also add a case with `Projector`, as well as cases with `Projector` and `Hermitian` inside a `Prod`?

File: tests/ops/functions/test_bind_new_parameters.py [hunk 5]
```diff
@@ -215,6 +226,45 @@ def test_hamiltonian(H, new_coeffs, expected_H):
...
+        ),
+    ),
+    (
+        qml.ops.LinearCombination(
+            [1.6, -1], [qml.Hermitian(np.array([[0.0, 1.0], [1.0, 0.0]]), wires=1), qml.PauliX(1)]
+        ),
+        [-1, np.array([[1.0, 1.0], [1.0, 1.0]]), 1.6],
+        qml.ops.LinearCombination(
+            [-1, 1.6], [qml.Hermitian(np.array([[1.0, 1.0], [1.0, 1.0]]), wires=1), qml.PauliX(1)]
+        ),
```
Comment (mudit2812) [inline tests/ops/functions/test_bind_new_parameters.py:248]: Same comment as above here

File: tests/ops/functions/test_bind_new_parameters.py (hunk 4 above)
Comment (mudit2812) [reply tests/ops/functions/test_bind_new_parameters.py:86]: Also it seems like the same thing is being tested twice, once in this test, and one in the new test you added. Probably a good idea to remove these test cases and just keep `test_linear_combination`.

**Review by mudit2812 — COMMENTED** [review]
Looks good to me! A little bit more testing needs to be added, I'm ready to approve after that.

File: tests/ops/functions/test_bind_new_parameters.py [hunk 6]
```diff
@@ -215,6 +226,45 @@ def test_hamiltonian(H, new_coeffs, expected_H):
...
+    TEST_BIND_LINEARCOMBINATION,
+)
+def test_linear_combination(H, new_coeffs, expected_H):
+    """Test that `bind_new_parameters` with `LinearCombination` returns a new
+    operator with the new parameters without mutating the original
+    operator."""
+    new_H = bind_new_parameters(H, new_coeffs)
+
+    assert qml.equal(new_H, expected_H)
+    assert new_H is not H
```
Comment (mudit2812) [inline tests/ops/functions/test_bind_new_parameters.py:264]: Can you also check that the `operands` and `ops` are not the same as the original operator's `operands` and `ops`?

Comment (mudit2812) [reply tests/ops/functions/test_bind_new_parameters.py:264]:
```suggestion
    assert new_H is not H
    assert all(no is not o for no, o in zip(new_op.operands, op.operands))
    assert all(no is not o for no, o in zip(new_op.ops, op.ops))
```

**Review by astralcai — APPROVED** [review]
Looks good

File: tests/ops/functions/test_bind_new_parameters.py (hunk 6 above)
Comment (Qottmann) [reply tests/ops/functions/test_bind_new_parameters.py:264]: > `assert all(no is not o for no, o in zip(new_op.ops, op.ops))`

For the latter I would have to make a copy of the `ops` in `bind_new_paramters`, but I actually dont think that is desired/necessary?

**Review by Qottmann — COMMENTED** [review]
thanks @mudit2812 for the suggestions with the tests. Added those now.
Not sure about checking that none of the ops should be the same instances (i.e. in cases of non-parametric ops like `qml.X(0)` I would deem it okay to have the same instance without creating new ones?)

**Comment by mudit2812** [comment]
> I would deem it okay to have the same instance without creating new ones? 

I agree, but `bind_new_parameters` always returns new instances regardless of whether the operator is parametric or not. This is also validated in tests for other composite ops. Not super necessary, but I'd rather stay consistent with other tests.

File: tests/ops/functions/test_bind_new_parameters.py (hunk 6 above)
Comment (mudit2812) [reply tests/ops/functions/test_bind_new_parameters.py:312]: Ah, I apologize, I forgot that we were not copying non-parametric ops. Feel free to revert the deepcopying change you made and disregard this comment about testing.

Comment (Qottmann) [reply tests/ops/functions/test_bind_new_parameters.py:312]: okay thanks! :)

File: tests/ops/functions/test_bind_new_parameters.py [hunk 7]
```diff
@@ -262,6 +310,8 @@ def test_linear_combination(H, new_coeffs, expected_H):
 
     assert qml.equal(new_H, expected_H)
     assert new_H is not H
+    assert all(no is not o for no, o in zip(new_H.operands, H.operands))
+    assert all(no is not o for no, o in zip(new_H.coeffs, H.coeffs))
```
Comment (mudit2812) [inline tests/ops/functions/test_bind_new_parameters.py:314]: Optional suggestion: Not sure if this is needed anymore after the above discussion. Especially if any of the new/old coefficients are the same (or are `==1`), then it also doesn't make much sense to test this.
```suggestion
```

**Review by mudit2812 — APPROVED** [review]

**Pull request merged**